src/Common/tests/Collections/IListTest.cs

```diff
@@ -0,0 +1,1692 @@
+                }
+                items = items.Push(items);
+            }
+            CollectionAssert.Equal(items, list);
+        }
+
+        [Fact]
+        public void InsertClearInsert()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+                return;
+            var items = GenerateItems(16);
+            var list = GetList(null);
+            list.InsertRange(items);
+            list.Clear();
+            items = GenerateItems(16);
+            list.InsertRange(items);
+            CollectionAssert.Equal(items, list);
+        }
+
+        [Fact]
+        public void InsertRemoveInsert()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+                return;
+            var items = GenerateItems(32);
+            var list = GetList(null);
+            list.InsertRange(items);
+            var tempItems = (object[]) items.Clone();
+            for (int i = 0; i < 16; i++)
+            {
+                list.RemoveAt(0);
+            }
+
+            Array.Copy(tempItems, 16, items, 0, 8);
+            Array.Copy(tempItems, 24, items, 24, 8);
+            tempItems = GenerateItems(16);
+            for (int i = 0; i < 16; i++)
+            {
+                list.Insert(8 + i, tempItems[i]);
+            }
+
+            Array.Copy(tempItems, 0, items, 8, 16);
+            CollectionAssert.Equal(items, list);
+        }
+
+        [Fact]
+        public void InsertRemoveAllInsert()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+                return;
+            var items = GenerateItems(32);
+            var list = GetList(null);
+            list.InsertRange(items);
+            for (int i = 0; i < 32; i++)
+            {
+                list.Remove(items[i]);
+            }
+            items = GenerateItems(32);
+            list.InsertRange(items);
+            CollectionAssert.Equal(items, list);
+        }
+
+        [Fact]
+        public void InsertAndAdd()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+                return;
+            var items = GenerateItems(48);
+            var list = GetList(null);
+            list.InsertRange(0, items, 0, 16);
+            list.AddRange(items, 16, 16);
+            list.InsertRange(32, items, 32, 16);
+            CollectionAssert.Equal(items, list);
+        }
+
+        [Fact]
+        public void InsertAndRemove()
+        {
+            if (ExpectedIsFixedSize || ExpectedIsReadOnly)
+                return;
+            var items = GenerateItems(32);
+            var tempItems = new object[33];
+            var list = GetList(items);
+            for (int i = 0; i < items.Length; i++)
+            {
+                var newObject = GenerateItem();
+                list.Insert(i, newObject);
+                Array.Copy(items, 0, tempItems, 0, i);
+                tempItems[i] = newObject;
+                Array.Copy(items, i, tempItems, i + 1, items.Length - i);
```

Comment:
I could remove the `O(N^2)` complexity using Enumerable.Concat if we need to. I don't think its necessary though because `n` is only 32.